### ISSUE TITLE
Room::disconnect() blocks and triggers reconnect/crash when offline

### ISSUE THREAD

**Issue by tomakor549** [livekit/client-sdk-cpp#222]
### LiveKit C++ SDK Version

v1.4.0

### Platform

Linux x86_64

### OS Version

Ubuntu 26

### Compiler & Version

g++ 15.2.0

### CMake Version

4.2.3

### How are you using the SDK?

Building from source (build.sh / build.cmd)

### Issue Description

Calling `Room::disconnect(DisconnectReason::ClientInitiated)` while the network is down blocks the calling thread. When connectivity returns, internal `rtc_engine` attempts reconnect, fails with `StateMismatch`, and the process can crash.

shutdown function part:

void LiveKitService::shutdown() {
  //...
  qCInfo(LiveKitLog) << "_livekit_audio_source.reset()...";
  _livekit_audio_source.reset();
  qCInfo(LiveKitLog) << "_livekit_video_source.reset()...";
  _livekit_video_source.reset();
  qCInfo(LiveKitLog) << "_delegate.reset()...";
  _delegate.reset();

  if (_livekit_initialized.exchange(false)) {
    qCInfo(LiveKitLog) << "_room.disconnect...";
    _room.disconnect(DisconnectReason::ClientInitiated);
  }
}

### Steps to Reproduce

1. Connect to a LiveKit room
2. Disable network (or unplug)
3. Call `_room.disconnect(DisconnectReason::ClientInitiated)`
4. Observe: thread blocks
5. Re-enable network
6. Disconnect unblocks, then reconnect errors appear; crash may follow

### Logs

```shell
##Example LOGS
[2026-07-28 19:22:34.077][LiveKitLog][INFO][53865][livekit::LiveKitService::shutdown:165]: _room.disconnect...
[2026-07-28T17:22:49Z ERROR livekit_api::signal_client::signal_stream] unhandled websocket message Err(Io(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }))
[2026-07-28T17:22:49Z ERROR livekit::rtc_engine] resuming connection failed: signal failure: server sent leave during reconnect: reason=StateMismatch, action=Reconnect
[2026-07-28T17:22:49Z ERROR livekit::rtc_engine] failed to reconnect to the livekit room
19:22:50: The command "..." terminated abnormally.
```

**Comment by alan-george-lk** [comment]
Thanks for flagging this -- will get back to you ASAP.

**Comment by tomakor549** [comment]
There is a similar problem with unpublishTrack. After internet disconnection, calling `unpublishTrack` blocks the calling thread indefinitely and never returns.

`videoTrack` is a `std::shared_ptr<livekit::LocalVideoTrack>` (local video track).

```cpp
try {
  if (auto participant = _room.localParticipant().lock()) {
    for (const auto& [sid, pub] : participant->trackPublications()) {
      if (pub->name() == videoTrack->name()) {
        participant->unpublishTrack(sid);  // ← hangs here after network drop
        break;
      }
    }
  }
} catch (const std::exception& e) {
  qCWarning(LiveKitLog()) << "Error during unpublishing track:" << e.what();
}

**Comment by alan-george-lk** [comment]
Just a quick follow up -- I am seeing the issue on some testing and a fix is definitely needed. Not sure when we can get this in, ideally in the next few weeks.

**Comment by lmengnr** [comment]
I may be seeing the same or a closely related issue on Windows x64 with LiveKit C++ SDK 1.5.0.

I reproduced the crash twice, although it is intermittent. Both times:

1. The application was connected to a room.
2. The internet connection went down.
3. Our application released its final reference to the room, causing room destruction and the implicit `Room::disconnect()`.
4. The process terminated with a fatal exception inside `livekit_ffi.dll`.

Visual Studio reported (for one of the events):

```text
Unhandled exception at 0x00007FF87EE15206 (livekit_ffi.dll) in N***n.exe.8856.dmp: Fatal program exit requested.
```

Windows Error Reporting recorded:

```text
Faulting module: livekit_ffi.dll
Exception code: 0xc0000409
Fault offset: 0x2235206
WER parameter P9: 7 (FAST_FAIL_FATAL_APP_EXIT)
```

Both crashes had the same `livekit_ffi.dll+0x2235206` fault offset.

The dump identified the exception thread as a `tokio-rt-worker`. At the same time, the main thread was destroying the LiveKit room -> The main-thread stack showed the application releasing its final room reference and entering LiveKit room destruction. I cannot share the application-specific frames publicly.

My current understanding is:

**Confirmed**

- The fatal exception occurred on a Tokio worker inside `livekit_ffi.dll`.
- The main thread was destroying/disconnecting the room at the same time.
- Both crashes reached the same offset inside `livekit_ffi.dll`.

**Strongly suspected**

- Room teardown overlapped with outstanding reconnect or event work inside LiveKit.

**Not yet confirmed**

- The exact LiveKit function, object, or invariant that caused the fatal exit.

Unfortunately, I did not have matching `livekit.pdb` or `livekit_ffi.pdb` files, so I could not resolve the internal LiveKit frames. This was an internally compiled Debug SDK build, but the PDB files were not included in our SDK package.

The failure seems related to this issue, although in my case the fatal exit occurred while the computer was still offline rather than after connectivity returned.

I will work on getting some more details for this if possible.

---

Disclosure: I took help from AI to analyze things and compose this report but I did verify the stack traces and the windows event viewer events myself.

**Comment by alan-george-lk** [comment]
@tomakor549 Following up on this -- I think my prior "reproduction" was actually a deadlock issue I introduced unrelated to this. I have a local tester app that allows me to programmatically kill the connection, and I'm not seeing the crash. I also noticed in your code report that you are calling `_delegate.reset()` prior to room disconnect. This is a red flag -- during disconnect the room will notify the delegate, so I am wondering if this is a use-after-free for your initial report. 

We can certainly harden the `room.cpp` code to better null-check the delegate, but I am curious if you adjust your example code if this still occurs?

**Comment by tomakor549** [comment]
@alan-george-lk  I put _delegate.reset() later but for this code _room.disconnect operation still block thread. log livekit::shutdown don't show. When I disconnect from the internet, the block on _room.disconnect remains in effect until I reconnect to the network. This happens even without first calling a reset for the delegate. The same with version 1.7.0. Unpublish track have the same problem.

```
void LiveKitService::shutdown() {
  qCInfo(LiveKitLog);
  //...

  qCInfo(LiveKitLog) << "_livekit_audio_source.reset()...";
  _livekit_audio_source.reset();
  qCInfo(LiveKitLog) << "_livekit_video_source.reset()...";
  _livekit_video_source.reset();

  if (_livekit_initialized.exchange(false)) {
    qCInfo(LiveKitLog) << "_room.disconnect...";
    _room.disconnect(DisconnectReason::ClientInitiated);
    qCInfo(LiveKitLog) << "livekit::shutdown...";
    livekit::shutdown();
  }

  qCInfo(LiveKitLog) << "_delegate.reset()...";
  _delegate.reset();
}
```

> [2026-08-24 13:07:34.390][LiveKitLog][INFO][45977][livekit::LiveKitService::shutdown:155]: _livekit_audio_source.reset()...
> [2026-08-24 13:07:34.390][LiveKitLog][INFO][45977][livekit::LiveKitService::shutdown:157]: _livekit_video_source.reset()...
> [2026-08-24 13:07:34.390][LiveKitLog][INFO][45977][livekit::LiveKitService::shutdown:161]: _room.disconnect...

I created a workaround for `unpublishTrack` because it also gets stuck on that line. In the logs, I have a sample timeout set to 5000 ms:

```
void LiveKitService::destroyVideoSource(std::shared_ptr<livekit::LocalVideoTrack>& videoTrack,
                                        std::shared_ptr<livekit::VideoSource>&     videoSource) {
  qCInfo(LiveKitLog());
  if (!videoTrack) {
    qCInfo(LiveKitLog()) << "Second video track does not exist or already disconnected.";
    return;
  }

  try {
    // videoTrack may not expose its SID directly, so locate it within the participant's published tracks.
    if (auto participant = _room.localParticipant().lock()) {
      for (const auto& [sid, pub] : participant->trackPublications()) {
        if (pub->name() == videoTrack->name()) {
          // Unpublish asynchronously with a timeout to prevent thread blocking.
          // Note: If connection drops suddenly, connectionState may briefly remain 'Connected',
          // causing unpublishTrack() to hang waiting for an ACK.
          auto              promise = std::make_shared<std::promise<void>>();
          std::future<void> future  = promise->get_future();

          // Pass variables by value to avoid dangling references if the detached thread outlives this scope
          std::thread([participant, sid, &promise]() {
            try {
              participant->unpublishTrack(sid);
              promise->set_value();
            } catch (...) {
              promise->set_exception(std::current_exception());
            }
          }).detach();

          if (future.wait_for(std::chrono::milliseconds(5000)) ==
              std::future_status::timeout) {
            qCWarning(LiveKitLog()) << "unpublishTrack timeout – skipping";
          } else {
            future.get();  // Re-throws exception from inside thread if unpublish failed with an error
            qCInfo(LiveKitLog()) << "Video track" << sid << "unpublished.";
          }
          break;
        }
      }
    }
  } catch (const std::exception& e) {
    qCWarning(LiveKitLog()) << "Error during unpublishing track:" << e.what();
  }

  videoTrack.reset();
  videoSource.reset();

  qCInfo(LiveKitLog()) << "Video source and track resources cleared.";
}
```

> [2026-08-24 13:19:57.381][LiveKitLog][INFO][51898][livekit::LiveKitService::destroyVideoSource:72]: 
> [2026-08-24 13:20:02.381][LiveKitLog][WARNING][51898][livekit::LiveKitService::destroyVideoSource:100]: unpublishTrack timeout – skipping
> [2026-08-24 13:20:02.381][LiveKitLog][INFO][51898][livekit::LiveKitService::destroyVideoSource:116]: Video source and track resources cleared.
> [2026-08-24 13:20:02.382][LiveKitLog][INFO][51898][livekit::LiveKitService::destroyVideoSource:72]: 
> [2026-08-24 13:20:02.382][LiveKitLog][INFO][51898][livekit::LiveKitService::destroyVideoSource:74]: Second video track does not exist or already disconnected.

These instances become active—they perform operations only after I connect again to the network.

**Comment by alan-george-lk** [comment]
Hey @tomakor549, appreciate your patience on this, juggling priorities internally. I'd like to first reproduce the original reported issue and then move to the data track one. I have created #247 which has a tester app. From the details you've mentioned, I've tried to mirror it as close as I can, but cannot get it to crash on Linux (working through some other CI issues on that branch but linux is good). Please review, and also feel free to open a PR against that one with edits that make it reproducible, sticking to LiveKit-only APIs.

Once we get it reproducing, then I can start to triage and move forward.

Note that I am basing this testing on the latest `main` which is SDK `v1.10.0`. I see you're on `v1.4.0`. There's a chance upstream Rust SDK version bumps have resolved this which is why I'm not seeing it, perhaps try an upgrade as well. Should be fully backwards compatible.

**Comment by tomakor549** [comment]
Thank you very much for looking into this. I've updated the SDK to version v1.10.0. The old crash from version v1.4.0 no longer occurs. However, the issue with threads getting blocked for "disconnect" and "unpublish" still persists. I'll try to run and edit a test to reproduce this issue in the mentioned PR.

**Comment by tomakor549** [comment]
@alan-george-lk I checked out the repo and analyzed the problem. It was related to the Room::shutdown(bool disconnect_ffi, DisconnectReason reason, bool notify_delegate) operation. The issue was with the line "FfiClient::instance().disconnectAsync(handle->get(), reason).get();" — this `get` call, without a network connection, was blocking the thread. It wouldn’t unblock until the connection was reestablished. I wrapped this operation in a separate `std::future` with a sample timeout of 5 seconds. After that timeout, I set the `shutdown_ok` operation to `false`. I'm not sure if this is safe, but I'm guessing that when closing connections, this operation will be canceled anyway? I also added an include for chrono. I built it locally and integrated it into my project; it closed the connection correctly, and I didn't have any problems opening subsequent ones either.

**I imagine there might be a similar issue with `unpublishTrack`**: 
```
//void LocalParticipant::unpublishTrack(const std::string& track_sid)
  auto fut = FfiClient::instance().unpublishTrackAsync(static_cast<std::uint64_t>(handle_id), track_sid,
                                                       /*stop_on_unpublish=*/true);
  fut.get();
```

Disconnect code snippet:
```
if (disconnect_ffi && claimed_disconnect && handle && handle->valid()) {
    try {
      auto future = FfiClient::instance().disconnectAsync(handle->get(), reason);
      constexpr auto kDisconnectTimeout = std::chrono::seconds(5);
      if (future.wait_for(kDisconnectTimeout) != std::future_status::ready) {
        LK_LOG_ERROR("Room shutdown: FFI disconnect timed out after {}s (continuing local shutdown)",
                     kDisconnectTimeout.count());
        shutdown_ok = false;
      } else {
        future.get();
      }
    } catch (const std::exception& e) {
      LK_LOG_ERROR("Room shutdown: FFI disconnect failed (continuing local shutdown): {}", e.what());
      shutdown_ok = false;
    } catch (...) {
      LK_LOG_ERROR("Room shutdown: FFI disconnect failed (continuing local shutdown): unknown exception");
      shutdown_ok = false;
    }
  }
```

**Comment by alan-george-lk** [comment]
@tomakor549 Glad to hear updating resolved the initial issue, I'll get back on this other one ASAP.
